bids/variables/variables.py

```diff
@@ -448,12 +448,31 @@ def resample(self, sampling_rate, inplace=False, kind='linear'):
         self.index = self._build_entity_index(self.run_info, sampling_rate)
 
         x = np.arange(n)
-        num = int(np.ceil(n * sampling_rate / old_sr))
+        # In Index we trust!
+        num = len(self.index)
+        # _build_entity_index computation above does it per run,
+        # which possibly provides a more stable solution and yadadada
+        num_computed = int(np.ceil(n * sampling_rate / old_sr))
+        num_diff = abs(num - num_computed)
+        if num_diff > 1:
+            raise RuntimeError(
+                "Our internal assumptions about resampling are too faulty "
+                "to continue.")
+        elif num_diff:
+            assert num_diff == 1  # the only way to get here
```

Comment:
Let's remove the assertions. They behave differently based on optimization level.

Reply:
> Let's remove the assertions.

I would say "let's add more assertions!" ;-)  They are good, and possibly a time/life saver if there is an inherent assumption in the code (e.g. that trial onsets are non-descending, that duplicate information about durations is congruent across different places, etc) which drives the code logic.  They allow to catch problematic usecases which invalidate the assumptions without silently providing incomplete or bogus results.

> They behave differently based on optimization level.

Yes, if you run via `python -O` assertion statements are simply ignored and no sane code "upstairs" should catch and treat AssertionError to guide its logic (we are, khe khe, to [get rid of those in datalad](https://github.com/datalad/datalad/pull/2832)].   In my experience I have never saw any notable performance boost by running with -O so I stopped using it (in PyMVPA we were somewhat meticulous about assertions and `__debug__` modes, but never saw pragmatic benefit). Or there is some other detrimental behavior of assertions which I don't know?  Do you often run code with any of the optimization levels for a real pragmatic benefit?

Reply:
I don't run with optimizations, but the point is that behavior can be inconsistent across installs. But you can check your assumptions in a way that will work the same regardless of optimization flags:

```Python
if not <assertion condition>:
    raise <relevant exception>
```

Reply:
Okay. I can live with assertions. I was under the impression they were generally disrecommended.